File: ake_online/ake_online_query.py
#!/usr/bin/python

#================================================================================#
from string import Template
#================================================================================#
class Query:
	def __init__(self):
		self.statement	= None
		self.result		= None
#================================================================================#
class Result:
	def __init__(self, columns=None, rows=None, count=0, rowcount=0, lastrowid=0):
		self.columns	= columns
		self.rows		= rows
		self.count		= count
		self.rowcount	= rowcount
		self.lastrowid	= lastrowid
#================================================================================#
class Recordset:
	def __init__(self): self.__records = []
	def empty(self): self.__records = []
	def add(self, record): self.__records.append(record)
	def iterate(self): return self.__records
	def insert(self):
		for record in self.__records: record.insert()
#================================================================================#
class Database:
	def __init__(self, database=None, username=None, password=None, host=None):
		self.__database		= database
		self.__username		= username
		self.__password		= password
		self.__host			= host
		self.__connection	= None
		self.__cursor		= None
		self.connect()
	#--------------------------------------#
	def connect(self):
		if(self.__database):
			if(self.__username):
				if(self.__password):
					if(self.__host):
						import mysql.connector
						self.__connection = mysql.connector.connect(database=self.__database, user=self.__username, password=self.__password, host=self.__host)
					else:
						import cx_Oracle
						self.__connection = cx_Oracle.connect(self.__username, self.__password, self.__database)
			else:
				import sqlite3
				self.__connection = sqlite3.connect(self.__database)
		self.cursor()
		
	def cursor(self): self.__cursor	= self.__connection.cursor()
	def commit(self): self.__connection.commit()
	def rollback(self): self.__connection.rollback()
	def close(self): self.__connection.close()
	#--------------------------------------#
	def execute(self, query):
		if(query.statement is not None):
			#print(query.statement)
			#result is a cursor object instance contains the returned records of select statement
			#None is the returned value in case of insert/update/delete.
			self.__cursor.execute(query.statement)
			rows=None
			count=None
			if(str((query.statement.strip())[:6]).lower()=="select"):
				rows = self.__cursor.fetchall()
				count = len(rows)
			#rowcount is readonly attribute
			#rowcount contains the count/number of the inserted/updated/deleted records/rows.
			#rowcount is -1 in case of rows/records select.
			rowcount = self.__cursor.rowcount

			#if(self.__cursor.lastrowid): lastrowid = self.__cursor.lastrowid
			columns = []
			#cursor.description returns a tuple of information describes each column in the table.
			#(name, type_code, display_size, internal_size, precision, scale, null_ok)
			#for index, column in enumerate(self.__cursor.description): columns.append(column[0])
			
			if(self.__cursor.description is not None): columns = [column[0] for column in self.__cursor.description]
			#query.result = Result(columns, rows, count, rowcount, lastrowid)
			query.result = Result(columns, rows, count, rowcount)
			return query
	#--------------------------------------#
        def executeScript(self, sqlScriptFileName):
                sqlScriptFile = open(sqlScriptFileName,'r')
                sql = sqlScriptFile.read()
                return self.__cursor.executescript(sql)
#================================================================================#
class ObjectRelationalMapper:
	def __init__(self): pass
	#--------------------------------------#
	def map(self, query, passedObject):
		object = passedObject
		if(query.result.rows): #if(query.result is not None):
			for row in query.result.rows:
				index = 0
				for fieldValue in row:
					# str() don't use # to map Null value to None field correctly. 
					setattr(object, query.result.columns[index], fieldValue)
					index += 1
				passedObject.mappedObjectsFromRecords.append(object)
				object = passedObject.__class__() #object = Record() #bug
#================================================================================#
class State:
	def __init__(self): self.current = None
#================================================================================#
class Record:
	database						= None
	orm								= ObjectRelationalMapper()
	# select, update and delete
	fieldValueTemplate				= Template('''${field}=${value}${separator}''')
	# insert
	field							= Template('''${field}, ''')
	value							= Template('''${value}, ''')
	fieldsValuesTemplate			= Template('''(${fields}) VALUES (${values})''')
	# ------
	selectStatementTemplate			= Template('''SELECT * FROM ${table} WHERE ${fieldsValues}''')
	deleteStatementTemplate			= Template('''DELETE FROM ${table} WHERE ${fieldsValues}''')
	insertStatementTemplate			= Template('''INSERT INTO ${table}${fieldsValues}''')
	updateStatementTemplate			= Template('''UPDATE ${table} SET ${fieldsValues} WHERE ${_fieldsValues_}''')
	#--------------------------------------#
	def __init__(self, statement=None, **kwargs):
		self.mappedObjectsFromRecords = []
		self.__state = State()
		if(kwargs is not None):
			for key, value in kwargs.items():
				setattr(self, key, value)
		self.__query = Query() # must be declared before self.queryObject(statement)
		if(statement is not None):
			self.__query.statement = statement
			#self. instead of Record. #change the static field self.__database for inherited children classes
			self.database.execute(self.__query)
			Record.orm.map(self.__query, self)
	#--------------------------------------#
	def __iter__(self): 
		self.__iterationIndex = 0
		self.__iterationBound = len(self.mappedObjectsFromRecords)
		return self
	#--------------------------------------#
	def __next__(self): #python 3 compatibility
		if(self.__iterationIndex < self.__iterationBound):
			currentItem = self.mappedObjectsFromRecords[self.__iterationIndex]
			self.__iterationIndex += 1
			return currentItem
		else:
			del(self.__iterationIndex) # to prevent using them as database's column
			del(self.__iterationBound) # to prevent using them as database's column
			raise StopIteration
	#--------------------------------------#
	def next(self): return self.__next__() #python 2 compatibility
	#--------------------------------------#
	def table(self): return self.__class__.__name__
	#--------------------------------------#
	def recordset(self): return self.mappedObjectsFromRecords #in PHP instead of #implements Iterator|IteratorAggregate
	#--------------------------------------#
	def fieldsEqualValues(self, separator=' AND ', fieldsValues = ''):
		separatorLength = len(separator)
		for attribute, value in self.__dict__.items(): #for a in dir(r): print(a)
			if(type(value) in [str, int]): # any other type will be execluded Class type, None type and others ...
				if(value != 'Null'): value = "'" + str(value) + "'"
				fieldsValues += Record.fieldValueTemplate.substitute({'field': attribute, 'value': value, 'separator': separator})
		if(fieldsValues): return fieldsValues[:-separatorLength]
		else: return None
	#--------------------------------------#
	def fieldsThenValues(self, fields='', values=''):
		for attribute, value in self.__dict__.items(): #for a in dir(r): print(a)
			if(type(value) in [str, int]): # any other type will be execluded Class type, None type and others ...
				if(value != 'Null'): value = "'" + str(value) + "'"
				fields += Record.field.substitute({'field': attribute})
				values += Record.value.substitute({'value': value})
		fields = fields[:-2]
		values = values[:-2]
		fieldsValues = Record.fieldsValuesTemplate.substitute({'fields': fields, 'values': values})
		if(fieldsValues): return fieldsValues
		else: return None
	#--------------------------------------#
	def __crd(self, template, fieldsValues, _fieldsValues_=None):
		if(fieldsValues is not None):
			self.__query.statement = template.substitute({'table': self.table(), 'fieldsValues': fieldsValues, '_fieldsValues_': _fieldsValues_})
			self.__state.current = None
			self.database.execute(self.__query)
			Record.orm.map(self.__query, self)
	#--------------------------------------#
	def startUpdate(self): self.__state.current = self.fieldsEqualValues()
	#--------------------------------------#
	def insert(self): self.__crd(Record.insertStatementTemplate, self.fieldsThenValues())
	def read(self)  : self.__crd(Record.selectStatementTemplate, self.fieldsEqualValues())
	def delete(self): self.__crd(Record.deleteStatementTemplate, self.fieldsEqualValues())
	def update(self):
		if(self.__state.current is not None):
			self.__crd(Record.updateStatementTemplate, self.fieldsEqualValues(', '), self.__state.current)
		else:
			print("The update doesn't start, no current state !")
	#--------------------------------------#
#================================================================================#

File: ake_online/ake_online_query.php
#!/usr/bin/php

<?php
header("Content-Type:text/plain");
#header("Content-Type:text/html");

#================================================================================#
class Template{
	private $__templateString;
	private $__string;
	public function __construct($string){
		$this->__templateString = $string;
	}
	public function substitute($keywords){
		$this->__string = $this->__templateString;
		foreach($keywords as $key => $value){
			$key = '${' . $key . '}';
			$this->__string = str_replace($key, $value, $this->__string);
		}
		return $this->__string;
	}
}
#================================================================================#
function println($string){echo($string . "\n");}
#================================================================================#
class Query{
	public function __construct(){
		$this->statement	= Null;
		$this->result		= Null;
	}
}
#================================================================================#
class Result{
	public function __construct($columns=Null, $rows=Null, $count=0, $rowcount=0, $lastrowid=0){
		$this->columns	= $columns;
		$this->rows		= $rows;
		$this->count	= $count;
		$this->rowcount	= $rowcount;
		$this->lastrowid = $lastrowid;
	}
}
#================================================================================#
class Recordset{
	private $__records;
	public function __construct(){
		$this->__records = []; #mapped objects from records
	}
	public function empty(){ $this->__records = []; }
	public function add($record){ array_push($this->__records, $record); }
	public function iterate(){ return $this->__records; }
}
#================================================================================#
class Database{
	private $__database;
	private $__username;
	private $__password;
	private $__host;
	private $__connection;
	private $__cursor;
	public function __construct($database=Null, $username=Null, $password=Null, $host=Null){
		$this->__database	= $database;
		$this->__username	= $username;
		$this->__password	= $password;
		$this->__host		= $host;
		$this->__connection	= Null;
		$this->__cursor		= Null;
		$this->connect();
	}
	#--------------------------------------#
	public function connect(){
		if($this->__database){
			if($this->__username){
				if($this->__password){
					if($this->__host){
						$this->__connection = new PDO("mysql:host=$this->__host;dbname=$this->__database", $this->__username, $this->__password);
					}else{
						
					}
				}
			}else{
				$this->__connection = new PDO($this->__database);
			}
		}
		$this->cursor();
	}
	public function cursor(){ $this->__cursor = $this->__connection;}
	public function beginTransaction(){ $this->__connection->beginTransaction(); }
	public function commit(){ $this->__connection->commit(); }
	public function rollback(){ $this->__connection->rollback(); }
	public function close(){ $this->__connection->close(); }
	#--------------------------------------#
	public function execute($query){
		if($query->statement != Null){
			#println($query->statement);
			$q = $this->__cursor->query($query->statement);
			$rows = Null;
			$count = Null;
			if(strtolower(substr($query->statement, 0, 6)) == "select"){
				$rows = $q->fetchAll(PDO::FETCH_ASSOC);
				$count = count($rows);
			}
			$columns = [];
			
			$query->result = new Result($columns, $rows, $count, $rowcount);
			return $query;
		}
	#--------------------------------------#
	}
}
#================================================================================#
class ObjectRelationalMapper{
	public function __construct(){}
	#--------------------------------------#
	public function map($query, $passedObject){
		$object = $passedObject;
		if($query->result->rows != Null){
			foreach($query->result->rows as $row){
				foreach($row as $column => $fieldValue){
					# str() don't use # to map Null value to None field correctly. 
					$object->{$column} = $fieldValue;
				}
				array_push($passedObject->mappedObjectsFromRecords, $object);
				$object = new Record();
			}
		}
	}
}
#================================================================================#
class State{
	public function __construct(){ $this->current = Null; }
}
#================================================================================#
class Record{
	public static $database = Null;
	private static $orm;
	# select, update and delete
	private static $fieldValueTemplate;
	# insert
	private static $field;
	private static $value;
	private static $fieldsValuesTemplate;
	# ------
	private static $selectStatementTemplate;
	private static $deleteStatementTemplate;
	private static $insertStatementTemplate;
	private static $updateStatementTemplate;
	
	#private $position;
	public $mappedObjectsFromRecords;
	private $__state;
	private $__query;
	#--------------------------------------#
	public function __construct($statement=Null){ #kwargs
		Record::$orm							= new ObjectRelationalMapper();
		# select, update and delete
		Record::$fieldValueTemplate				= new Template('${field}=${value}${separator}');
		# insert
		Record::$field							= new Template('${field}, ');
		Record::$value							= new Template('${value}, ');
		Record::$fieldsValuesTemplate			= new Template('(${fields}) VALUES (${values})');
		# ------
		Record::$selectStatementTemplate		= new Template('SELECT * FROM ${table} WHERE ${fieldsValues}');
		Record::$deleteStatementTemplate		= new Template('DELETE FROM ${table} WHERE ${fieldsValues}');
		Record::$insertStatementTemplate		= new Template('INSERT INTO ${table}${fieldsValues}');
		Record::$updateStatementTemplate		= new Template('UPDATE ${table} SET ${fieldsValues} WHERE ${_fieldsValues_}');

		#$this->position = 0;
		$this->mappedObjectsFromRecords = [];
		$this->__state = new State();
		#if(kwargs is not None):
		#	for key, value in kwargs.items():
		#		setattr(self, key, value)
		$this->__query = new Query(); # must be declared before self.queryObject(statement)
		if($statement != Null){
			$this->__query->statement = $statement;
			Record::$database->execute($this->__query);
			Record::$orm->map($this->__query, $this);
		}
	}
	#--------------------------------------#
	public function recordset(){ return $this->mappedObjectsFromRecords; } #in PHP instead of #implements Iterator|IteratorAggregate
	#--------------------------------------#
	public function table(){ return get_class($this); }
	#--------------------------------------#
	public function fieldsNames(){
		$fieldsNames = [];
		$fieldsEqual = '';
		$fields = '';
		foreach($this as $attribute => $value){ #for a in dir(r): println(a)
			if(in_array(gettype($value) , [string, integer], TRUE)){ # any other type will be execluded Class type, None type and others ...
				#if($value != 'Null'){ $value = "'" . (string)$value . "'"; }
				array_push($fieldsNames, $attribute);
				$fieldsEqual .= $attribute . '=?, ';
				$fields .= $attribute . ', ';
			}
		}
		println($fieldsEqual);
		println($fields);
		return $fieldsNames;
	}
	#--------------------------------------#
	public function filedsValues($filedsNames){
		$fieldsValues = [];
		foreach($fieldsNames as $fieldName){ array_push($fieldsValues, $this[$fieldName]); }
		return $filedsValues;
	}
	#--------------------------------------#
	public function fieldsEqualValues($separator=' AND ', $fieldsValues = ''){
		$separatorLength = strlen($separator);
		foreach($this as $attribute => $value){ #for a in dir(r): println(a)
			if(in_array(gettype($value) , [string, integer], TRUE)){ # any other type will be execluded Class type, None type and others ...
				if($value != 'Null'){ $value = "'" . (string)$value . "'"; }
				$fieldsValues .= Record::$fieldValueTemplate->substitute(['field' => $attribute, 'value' => $value, 'separator' => $separator]);
			}
		}
		if($fieldsValues != Null){ return substr($fieldsValues, 0, -$separatorLength); }
		else{ return Null; }
	}
	#--------------------------------------#
	public function fieldsThenValues($fields='', $values=''){
		foreach($this as $attribute => $value){ #for a in dir(r): println(a)
			if(in_array(gettype($value) , [string, integer], TRUE)){ # any other type will be execluded Class type, None type and others ...
				if($value != 'Null'){ $value = "'" . (string)$value . "'"; }
				$fields .= Record::$field->substitute(['field' => $attribute]);
				$values .= Record::$value->substitute(['value' => $value]);
			}
		}
		$fields = substr($fields, 0, -2);
		$values = substr($values, 0, -2);
		$fieldsValues = Record::$fieldsValuesTemplate->substitute(['fields' => $fields, 'values' => $values]);
		if($fieldsValues != Null){ return $fieldsValues; }
		else{ return Null; }
	}
	#--------------------------------------#
	private function __crd($template, $fieldsValues, $_fieldsValues_=Null){
		if($fieldsValues != Null){
			$this->__query->statement = $template->substitute(['table' => $this->table(), 'fieldsValues' => $fieldsValues, '_filedsValues' => $_fieldsValues_]);
		}
		$this->__state->current = Null;
		Record::$database->execute($this->__query);
		Record::$orm->map($this->__query, $this);
	}
	#--------------------------------------#
	public function startUpdate(){ $this->__state->current = $this->fieldsEqualValues(); }
	#--------------------------------------#
	public function insert(){ $this->__crd(Record::$insertStatementTemplate, $this->fieldsThenValues()); }
	public function read(){	$this->__crd(Record::$selectStatementTemplate, $this->fieldsEqualValues()); }
	public function delete(){ $this->__crd(Record::$deleteStatementTemplate, $this->fieldsEqualValues()); }
	public function update(){
		if($this->__state->current != Null){
			$this->__crd(Record::$updateStatementTemplate, $this->fieldsEqualValues(', '), $this->__state->current);
		}else{
			println("The update doesn't start, no current state !");
		}			
	}
	#--------------------------------------#
}
#================================================================================#
?>